improvement(sct-config): convert append_scylla_yaml to dict

dimakr opened the pull request:
Currently the 'append_scylla_yaml' attribute is defined in test configurations as a multiline string representation of a dict. And is processed correspondingly in SCT.
The downside of it is if there are several test configurations for a test run that have 'append_scylla_yaml' attribute in them, SCTConfiguration overwrites the attribute value for each loaded config. As a result the last test configuration 'wins', in terms of that only its value is added to scylla.yaml.

The change converts value of all 'append_scylla_yaml' attribute occurrences in configs to a dict, so that it is natively merged by anyconfig. It also adjusts places where 'append_scylla_yaml' is used in the code, to process it
as a dict, not a string.

Related task: https://github.com/scylladb/scylla-cluster-tests/issues/7364

### Testing
- [["configurations/audit.yaml","test-cases/features/stop-compaction.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/sct-debug/51/) this is to ensure that 'append_scylla_yaml' values from different configs are merged
- [["test-cases/longevity/longevity-topology-changes-3h.yaml","configurations/raft/enable_raft_experimental.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-AbortRepairMonkey-docker-test/37/) - experimental_features attribute is supported as a standalone attribute in test config, but we had a number of instances where it was a part of append_scylla_yaml. This build is to ensure that experimental_features attribute is correctly merged to scylla.yaml, after moving out of append_scylla_yaml.
- [["test-cases/artifacts/ubuntu2004-fips.yaml", "configurations/local-ear.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/artifacts-ubuntu2004-fips-test/) this is to ensure that when there were nested subattributes in string version of append_scylla_yaml, after conversion to dict they are correctly merged and added to the final scylla.yaml.
- [["test-cases/longevity/longevity-encryption-at-rest-200GB-6h.yaml", "configurations/kms-ear.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/append_scylla_yaml_test/2/) - this is to ensure that KMS logic is not affected, as its attributes was changed in this commit.

### PR pre-checks (self review)
- [ ] I added the relevant `backport` labels
- [X] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

dimakr commented:
There is a failed unit tests - unit_tests/test_base_version.py::TestBaseVersion::test_2021_1_with_centos8.
But I am not sure that this is something related to the change in question.

fruch commented on sdcm/sct_config.py:
```diff
@@ -1678,13 +1678,36 @@ def __init__(self):
+            """
+            if key == 'append_scylla_yaml':
+                if key in self and key in other:
+                    self[key] += other[key]
```
I'm sure pure string append would be enough in this case....

maybe we need to parse it from yaml, and then merge and turn it back to string ?

what happens if we'll have multiple values ?

i.e. keep in mind `append_scylla_yaml` is currently defined as string, that holds a yaml, and not part of the yaml

fruch replied:
regardless I think this specific change might need a unit test or a few of them
proving how it would work

dimakr replied:
>I'm sure pure string append would be enough in this case....

Not sure, maybe I'm missing something from the comment.
The `self[key] += other[key]` exactly appends value of `append_scylla_yaml` from one config (config that is being merged) to the value of the other (config to which merge is performed)

dimakr replied:
>regardless I think this specific change might need a unit test or a few of them
proving how it would work

I'll unit test. Test run on real test configurations show that merging the value works as expected.

fruch replied:
build a test that both config file has the same key in `append_scylla_yaml`, and you'll get a broken yaml inside of it

config1
```
append_scylla_yaml: |
  enable_sstables_mc_format: true
  enable_sstables_md_format: true
```

config2
```
append_scylla_yaml: |
  enable_sstables_mc_format: false
  enable_sstables_md_format: false
```

the outcome would be:

```
append_scylla_yaml: |
  enable_sstables_mc_format: true
  enable_sstables_md_format: true
  enable_sstables_mc_format: false
  enable_sstables_md_format: false
```

wouldn't it ?

fruch replied:
we need the ability to overwrite things inside the yaml **IN** `append_scylla_yaml` value

fruch replied:
we can change the definition of `append_scylla_yaml` to stop being a string, and make it a first class yaml structure.
but we'll need to carefully attend to the 68 usages of it while doing so. (it's not a very big change, but need to be careful)

fruch replied:
@soyacz saying the same https://github.com/scylladb/scylla-cluster-tests/pull/7554#issuecomment-2146955643

with with the exact details and example of other config option handling the same situation.

dimakr replied:
Ok, so we don't want to be able to merge values of `append_scylla_yaml` attribute as they are used now in the STC.
But to rework how this attribute is used and get rid of defining its value as a multiline string everywhere in SCT?

dimakr replied:
got it, will be back soon

fruch commented:
> There is a failed unit test - unit_tests/test_base_version.py::TestBaseVersion::test_2021_1_with_centos8. But I am not sure that this is something related to the change in question.

if that test keep failing, we can drop it.
it's point to a very old release folder, that might have been deleted.

even that the failure seem like some networking issue

soyacz commented:
@dimakr did you try replacing type of `append_scylla_yaml` sct config from `str` to `dict_or_str` so it behaves the same as `stress_image` which support override out of the box?

dimakr commented:
I converted `append_scylla_yaml` values in test configs from multiline string to dict values.
Also adjusted few places in code that previously were loading the string value of the attribute.

I executed few builds that use test configurations with `append_scylla_yaml` - there are passed and failed runs, which is expected as such test configurations fail in other jobs.
But in the scope of the current change - to ensure that merging of `append_scylla_yaml` values works - they do the job:
- [["configurations/audit.yaml","test-cases/features/stop-compaction.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/sct-debug/51/) this one is to ensure that 'append_scylla_yaml' values from different configs are merged
- [["test-cases/longevity/longevity-topology-changes-3h.yaml","configurations/raft/enable_raft_experimental.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-AbortRepairMonkey-docker-test/37/). I see that `experimental_features` attribute is supported as a standalone attribute in test config, but we had a number of instances where it was a part of  `append_scylla_yaml`. This build is to ensure that `experimental_features` attribute is correctly merged to scylla.yaml, after moving out of  `append_scylla_yaml`.
- [["test-cases/artifacts/ubuntu2004-fips.yaml", "configurations/local-ear.yaml"] configs](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/artifacts-ubuntu2004-fips-test/) just another run to ensure that when there were nested subattributes in string version of `append_scylla_yaml`, after conversion to dict they are correctly merged and added to the final scylla.yaml

@soyacz @fruch

fruch commented on configurations/audit.yaml:
```diff
@@ -1,5 +1,5 @@
-  audit_tables: "keyspace1.standard1,scylla_bench.test"
-  audit_keyspaces: "keyspace1,scylla_bench"
+append_scylla_yaml:
+  audit: 'syslog'
```
our of curiosity, when move from double quote to single ?

dimakr replied:
Single quotes are always safer in yaml, so that there is no need to escape spec. characters (like `\` or `"`, etc.) if they are introduced to string values in future.

soyacz replied:
I agree, e.g. double quotes in `defaults/manager_persistent_snapshots.yaml` caused us quite a lot of troubles - as are included in parsed dict.

fruch commented on test-cases/jepsen/jepsen_with_raft.yaml:
```diff
@@ -20,6 +20,5 @@ jepsen_test_count: 5
-append_scylla_yaml: |
-  experimental_features:
-    - raft
+experimental_features:
```
this job need to be delete... there's no such flag anymore

dimakr replied:
Only `experimental_features` attribute is to be deleted, or the whole test config + its dedicated jenkins job?

fruch commented on test-cases/performance/perf-regression-throughput-cdc-mixed-poll-batching.yaml:
```diff
@@ -24,9 +24,8 @@ store_perf_results: true
-append_scylla_yaml: |
-  experimental_features:
-    - cdc
+experimental_features:
```
those needs to be remove

fruch replied:
cdc isn't experimental for quite some time

fruch commented on test-cases/performance/perf-regression-write-throughput-cdc.yaml:
```diff
@@ -17,9 +17,8 @@ store_perf_results: true
-append_scylla_yaml: |
-  experimental_features:
-    - cdc
+experimental_features:
```
those needs to be remove 

cdc isn't experimental for quite some time

fruch reviewed (approved):
beside some cleanups, which are critical 

this looks version, and would improve the usability and readability of sct configuration

soyacz commented on configurations/tablets-initial-32.yaml:
```diff
@@ -1,5 +1,5 @@
-  experimental_features:
-    - tablets
-    - consistent-topology-changes
+experimental_features:
+  - tablets
+  - consistent-topology-changes
```
btw. recently these got removed from experimental.
To avoid hassle with merge conflicts later I'd propose to switch it to `enable_tablets: true` right away

dimakr replied:
Maybe it's better to do both changes - cleanup of usages of these 2 features as experimental and add `enable_tablets` flag to scylla.yaml - in a separate PR?
For `consistent-topology-changes` there is a test config, that sets only this option - enables as experimental feature - `consistent-topology-changes`. That test config is used in 31 places in the SCT.
For `tablets` - it is also used in 2 test configs that do only that - enable it as experimental feature. And these 2 configs are used in 11 places in SCT.

soyacz replied:
yes, thinking again we should split it to separate PR as it will also require change in `ScyllaYaml` - too much for this change.

soyacz commented:
I propose to run a test with KMS - @dimakr please run one test with Scylla enterprise so it's enabled and we see it's working.

vponomaryov reviewed (approved):
LGTM.

Agree that would be good to see an enterprise Scylla with KMS be tested.

fruch commented:
@dimakr did you test it with enterprise ?

dimakr commented:
> @dimakr did you test it with enterprise ?

Still in progress. Jon was killed due to spot instances.

fruch commented:
> > @dimakr did you test it with enterprise ?
> 
> @fruch Still in progress. Job was killed due to spot instances.

@dimakr 

I'm guessing this is the job running 
https://argus.scylladb.com/test/d86e92f6-f70b-4680-9882-67352cf5b0de/runs?additionalRuns[]=f8a7dbe1-b28b-4fb2-a98d-df2e392bf633

I would recommend putting a link to it, on the description (even when it's ongoing)

also please update the PR description, i.e. remove the part under testing that refer to examples of string concatenation, it would be confusing to who gonna read it later on

dimakr commented:
> I'm guessing this is the job running [https://argus.scylladb.com/test/d86e92f6-f70b-4680-9882-67352cf5b0de/runs?
Yes, this is the job. At the end I had to use on_demand instances as with spot ones there were always capacity related interruptions.
This job failed. But it failed with the same error as it had in the past, e.g. https://jenkins.scylladb.com/view/enterprise-2024.1/job/enterprise-2024.1/job/SCT_Enterprise_Features/job/EncryptionAtRest/job/EaR-longevity-kms-200gb-6h-test/2/.
Can it be considered that changes in attributes merging around KMS attributes haven't affected that logic?
@fruch @soyacz

soyacz commented:
> > I'm guessing this is the job running [https://argus.scylladb.com/test/d86e92f6-f70b-4680-9882-67352cf5b0de/runs?
> > Yes, this is the job. At the end I had to use on_demand instances as with spot ones there were always capacity related interruptions.
> > This job failed. But it failed with the same error as it had in the past, e.g. https://jenkins.scylladb.com/view/enterprise-2024.1/job/enterprise-2024.1/job/SCT_Enterprise_Features/job/EncryptionAtRest/job/EaR-longevity-kms-200gb-6h-test/2/.
> > Can it be considered that changes in attributes merging around KMS attributes haven't affected that logic?
> > @fruch @soyacz

You can  download db logs and see if related stuff is there in scylla.yaml files. Also can look if keyspaces were altered to be encrypted (possibly by searching for `ALTER ...` in sct log.

fruch commented:
> > I'm guessing this is the job running [https://argus.scylladb.com/test/d86e92f6-f70b-4680-9882-67352cf5b0de/runs?
> > Yes, this is the job. At the end I had to use on_demand instances as with spot ones there were always capacity related interruptions.

any job that run more than 3-4 hours, spot would be problematic for it, we are using instance which has quite high demand (some of it by us...) 

> > This job failed. But it failed with the same error as it had in the past, e.g. https://jenkins.scylladb.com/view/enterprise-2024.1/job/enterprise-2024.1/job/SCT_Enterprise_Features/job/EncryptionAtRest/job/EaR-longevity-kms-200gb-6h-test/2/.
> > Can it be considered that changes in attributes merging around KMS attributes haven't affected that logic?
> > @fruch @soyacz
